central/baseimage/datastore/datastore_impl.go

```diff
@@ -0,0 +1,136 @@
+package datastore
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/jackc/pgx/v5"
+	"github.com/pkg/errors"
+	baseImageStore "github.com/stackrox/rox/central/baseimage/store/postgres"
+	"github.com/stackrox/rox/generated/storage"
+	"github.com/stackrox/rox/pkg/postgres"
+	"github.com/stackrox/rox/pkg/uuid"
+)
+
+type datastoreImpl struct {
+	storage baseImageStore.Store
+	db      postgres.DB
+}
+
+const (
+	baseImagesTable = "base_images"
+	// The 'firstlayerdigest' column is indexed.
+	listByFirstLayerQuery = "SELECT id FROM " + baseImagesTable + " WHERE firstlayerdigest = $1"
+	// Query to find the ID by manifest digest
+	getByManifestDigestQuery = "SELECT id FROM " + baseImagesTable + " WHERE manifestdigest = $1"
+	upsertChunkSize          = 100
+)
+
+// New creates a new DataStore instance backed by PostgreSQL.
+func New(store baseImageStore.Store, db postgres.DB) DataStore {
+	return &datastoreImpl{
+		storage: store,
+		db:      db,
+	}
+}
+
+func (ds *datastoreImpl) UpsertImage(ctx context.Context, image *storage.BaseImage, digests []string) error {
+	layers(image, digests)
+	return ds.storage.Upsert(ctx, image)
+}
+
+func (ds *datastoreImpl) UpsertImages(
+	ctx context.Context,
+	imagesWithLayer map[*storage.BaseImage][]string,
+) error {
+	batch := make([]*storage.BaseImage, 0, len(imagesWithLayer))
+	for img, digests := range imagesWithLayer {
+		// Attach layer digests to the image before sending to storage.
+		layers(img, digests)
+		batch = append(batch, img)
+	}
+
+	if len(batch) <= upsertChunkSize {
+		return ds.storage.UpsertMany(ctx, batch)
+	}
+
+	// Process in chunks. This reduces pressure on storage and avoids parameter/time limits.
+	for start := 0; start < len(batch); start += upsertChunkSize {
+		end := start + upsertChunkSize
+		if end > len(batch) {
+			end = len(batch)
+		}
+
+		if err := ds.storage.UpsertMany(ctx, batch[start:end]); err != nil {
+			return fmt.Errorf("upsert images chunk [%d:%d]: %w", start, end, err)
+		}
```

Comment:
Do we need a transaction? My first reaction is "no", because partial updates here (in case a failure happens in a batch `start > 0`) would be OK. But I want to ask you the question anyway to make sure.

Reply:
We tolerates some images being upserted while others fail and avoid locks. So I think we should be ok with no transaction.

Reply:
That's OK. Make sure this contract is explicit in the `UpsertImages` -- I think typically callers would assume this is a "all or nothing" operation, and might get confused that returning an error here resulted in partial writes. Not very savvy on the datastore interfaces -- so, @c-du do you have any suggestions?